File: solutions/Poly-Knapsack.cpp
int solve(vector<int> &weights, vector<int> &values, int capacity)
{
    // unbounded knapsack
    vector<int> dp(capacity + 1);
    for (int cap = 0; cap <= capacity; cap++)
    {
        for (int item = 0; item < weights.size(); item++)
        {
            if (weights[item] <= cap)
                dp[cap] = max(dp[cap], dp[cap - weights[item]] + values[item]);
        }
    }

    return dp[capacity];
}

File: solutions/Poly-Knapsack.py
import functools


class Solution:
    def solve(self, weights, values, capacity):
        @functools.cache
        def max_value(cap):
            return max(
                [value + max_value(cap - weight) for value, weight in zip(values, weights) if weight <= cap],
                default=0,
            )

        return max_value(capacity)
